feg/gateway/services/n7_n40_proxy/notify/notify_handler.go

```diff
@@ -0,0 +1,206 @@
+//   Example: http://magma-feg.magma.com/sm-policy-control/v1/MTIzNDU2Nzg5MDsxMjM0NQo=/update
+//   Here notifyRoot = http://magma-feg.magma.com/sm-policy-control/v1
+//        encodedSessionId = MTIzNDU2Nzg5MDsxMjM0NQo= (Session-Id is urlencoded)
+//        operation = update
+// This notification url is send to PCF in the SmPolicyCreate request
+func (handler *NotificationHandler) RegisterHandlers() error {
+	urlDef, err := url.ParseRequestURI(handler.config.NotifyApiRoot)
+	if err != nil {
+		return fmt.Errorf("error parsing notify api root - %s", err)
+	}
+	basePath := fmt.Sprintf("%s/:%s", urlDef.Path, EncodedSessionId)
+	// Register the handlers
+	handler.notifyServer.POST(basePath+"/update", handler.postSmPolicyUpdateNotification)
+	handler.notifyServer.POST(basePath+"/terminate", handler.postSmPolicyTerminateNotification)
+	return nil
+}
+
+func (handler *NotificationHandler) Start() (string, error) {
+	errChan := make(chan error)
+	addr, err := net.ResolveTCPAddr("tcp", handler.config.LocalAddr)
+	if err != nil {
+		return "", err
+	}
+
+	go func() {
+		err = handler.notifyServer.Start(addr.String())
+		if err != nil {
+			errChan <- err
+		}
+	}()
+
+	return sbi.WaitForSbiServerAndGetAddr(handler.notifyServer, errChan)
+}
+
+func (handler *NotificationHandler) Shutdown() {
+	ctx, cancel := context.WithTimeout(context.Background(), ShutdownTimeout)
+	defer cancel()
+	handler.notifyServer.Shutdown(ctx)
+}
+
+// postSmPolicyUpdateNotification handles SM policy update notification request from PCF.
+func (handler *NotificationHandler) postSmPolicyUpdateNotification(ctx echo.Context) error {
+	client, err := relay.GetSessionProxyResponderClient(handler.cloudRegistry)
+	if err != nil {
+		glog.Errorf("postSmPolicyUpdateNotification failed to get SessionProxyResponderClient: %s", err)
+		return fmt.Errorf("internal server error")
+	}
+	defer client.Close()
+
+	var smUpdateNotify n7_client.SmPolicyNotification
+	err = ctx.Bind(&smUpdateNotify)
+	if err != nil {
+		err = fmt.Errorf("invalid SmPolicyNotification received: %s", err)
+		glog.Errorf("postSmPolicyUpdateNotification: %s", err)
+		return echo.NewHTTPError(http.StatusBadRequest, err)
+	}
+	sessionId, imsi, err := getSessionIdAndIMSI(ctx.Param(EncodedSessionId))
+	if err != nil {
+		glog.Errorf("postSmPolicyUpdateNotification unable to fetch session-id for UpdateNotify - %s", err)
+		return echo.NewHTTPError(http.StatusBadRequest, err)
+	}
+
+	policyReauthProto := n7.GetPolicyReauthRequestProto(sessionId, imsi, smUpdateNotify.SmPolicyDecision)
+	ans, err := client.PolicyReAuth(context.Background(), policyReauthProto)
+	if err != nil {
+		glog.Errorf("Error relaying N7 policy reauth request to gateway: %s", err)
+		return fmt.Errorf("internal server error")
+	}
+	partialSuccessReport := n7.GetPartialSuccessReportN7(ans)
+	if partialSuccessReport != nil {
+		ctx.JSON(http.StatusOK, partialSuccessReport)
+	}
+	return ctx.NoContent(http.StatusOK)
```

Comment:
This function seems to need more consistency on the returns:  either we return http errors, or fplain errors or ctx errors. 

why are you responding with this code?  Isn't nil enough

Reply:
ok now it make more sense. Before it was confusing what it was being returned in case of partial success